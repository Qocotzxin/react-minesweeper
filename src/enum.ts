export enum Piece {
  uncovered = "â–¡",
  bomb = "*",
}

export enum Message {
  lose = "You lose",
  ok = "OK",
  win = "You win",
}
