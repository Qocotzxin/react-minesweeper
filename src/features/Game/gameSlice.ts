import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Message } from "../../enum";

export interface GameState {
  map: string[];
  isReady: boolean;
  message: Message;
  isLoadingBoard: boolean;
  difficulty: number;
}

const initialState: GameState = {
  map: [],
  isReady: false,
  message: Message.ok,
  isLoadingBoard: false,
  difficulty: 1,
};

export const gameSlice = createSlice({
  name: "game",
  initialState,
  reducers: {
    startGame(state, action: PayloadAction<number | null>) {
      state.isLoadingBoard = true;
      state.message = Message.ok;
      state.difficulty = action.payload || 1;
    },
    updateBoard(state, action: PayloadAction<string>) {
      if (state.isLoadingBoard) {
        state.isLoadingBoard = false;
      }
      state.map = action.payload.split("map:")[1].split("\n").filter(Boolean);
    },
    updateMessage(state, action: PayloadAction<string>) {
      state.message = action.payload.split("open:")[1].trim() as Message;
    },
    confirmSocketIsReady(state) {
      state.isReady = true;
    },
  },
});

export const { startGame, updateBoard, confirmSocketIsReady, updateMessage } =
  gameSlice.actions;

export default gameSlice.reducer;
