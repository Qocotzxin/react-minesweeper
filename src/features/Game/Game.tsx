import { Box, LinearProgress, Stack, Typography } from "@mui/material";
import React, { useCallback } from "react";
import { useSelector } from "react-redux";
import { useAppDispatch } from "../../app/hooks";
import { RootState } from "../../app/store";
import { Board, LevelSelector, ResultMessage } from "../../components";
import { Message } from "../../enum";
import { selectCoordinates, startGame } from "./gameSlice";

export function Game() {
  const dispatch = useAppDispatch();
  const boardMap = useSelector((state: RootState) => state.game.map);
  const isReady = useSelector((state: RootState) => state.game.isReady);
  const message = useSelector((state: RootState) => state.game.message);
  const difficulty = useSelector((state: RootState) => state.game.difficulty);
  const isLoadingBoard = useSelector(
    (state: RootState) => state.game.isLoadingBoard
  );

  const onStart = useCallback(
    (levelDifficulty) => dispatch(startGame(levelDifficulty)),
    [dispatch]
  );

  const onBoardClick = useCallback(
    (coords) => dispatch(selectCoordinates(coords)),
    [dispatch]
  );

  return (
    <Stack
      spacing={8}
      width="100vw"
      height="100vh"
      bgcolor="#171923"
      color="whitesmoke"
      p="4rem"
    >
      <Stack width="100%" alignItems="center" justifyContent="center">
        <Typography component="h1" fontSize="2rem">
          Welcome to Minesweeper!
        </Typography>
        <LevelSelector onStart={onStart} isReady={isReady} />
        {isLoadingBoard ? (
          <Box width="100%">
            <LinearProgress color="secondary" />
          </Box>
        ) : (
          <Board
            boardMap={boardMap}
            onClick={onBoardClick}
            difficulty={difficulty}
          />
        )}
        {message !== Message.ok && <ResultMessage message={message} />}
      </Stack>
    </Stack>
  );
}
