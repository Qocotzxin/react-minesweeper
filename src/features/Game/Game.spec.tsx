import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { FC } from "react";
import { Provider } from "react-redux";
import { Message } from "../../enum";
import tileReducer from "../Tiles/tileSlice";
import { Game } from "./Game";
import gameReducer, { GameState } from "./gameSlice";

interface WrapperProps {
  preloadedState?: GameState;
}

const defaultGameState = {
  isLoadingBoard: false,
  map: [],
  isReady: true,
  message: Message.ok,
  difficulty: 1,
};

const defaultTileState = {
  availableFlags: 10,
  flaggedTiles: [],
};

const Wrapper: FC<WrapperProps> = ({ preloadedState = defaultGameState }) => {
  return (
    <Provider
      store={configureStore({
        reducer: { game: gameReducer, tile: tileReducer },
        preloadedState: { game: preloadedState, tile: defaultTileState },
      })}
    >
      <Game />
    </Provider>
  );
};

describe("Game", () => {
  describe("Snapshot", () => {
    it("Should render correctly and match snapshot.", () => {
      const { container } = render(<Wrapper />);
      expect(container).toMatchSnapshot();
    });
  });

  describe("Behavior.", () => {
    it("Should render a progress bar when the board is still loading.", () => {
      const state = {
        isLoadingBoard: true,
        map: [],
        isReady: false,
        message: Message.ok,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      expect(screen.getAllByRole("progressbar").length).toBeGreaterThan(0);
      expect(screen.queryByTestId("Board")).not.toBeInTheDocument();
    });

    it("Should render the board when its already loaded.", () => {
      const state = {
        isLoadingBoard: false,
        map: [],
        isReady: true,
        message: Message.ok,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      expect(screen.getByTestId("Board")).toBeInTheDocument();
      expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
    });

    it("Should render the message component when message is different from 'OK'.", () => {
      const state = {
        isLoadingBoard: false,
        map: [],
        isReady: true,
        message: Message.lose,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      expect(screen.getByTestId("ResultMessage")).toBeInTheDocument();
    });

    it("Should NOT render the message component when message is 'OK'.", () => {
      const state = {
        isLoadingBoard: false,
        map: [],
        isReady: true,
        message: Message.ok,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      expect(screen.queryByTestId("ResultMessage")).not.toBeInTheDocument();
    });

    it("Should NOT render the message component when message is 'OK'.", () => {
      const state = {
        isLoadingBoard: false,
        map: [],
        isReady: true,
        message: Message.ok,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      expect(screen.queryByTestId("ResultMessage")).not.toBeInTheDocument();
    });

    it("Should update the isLoadingBoard and message states when start game button is clicked.", () => {
      const state = {
        isLoadingBoard: false,
        map: [],
        isReady: true,
        message: Message.win,
        difficulty: 1,
      };
      render(<Wrapper preloadedState={state} />);
      fireEvent.click(screen.getByRole("button"));
      expect(screen.getAllByRole("progressbar").length).toBeGreaterThan(0);
      expect(screen.queryByTestId("ResultMessage")).not.toBeInTheDocument();
    });
  });
});
