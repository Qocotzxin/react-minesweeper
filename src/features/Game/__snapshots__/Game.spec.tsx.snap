// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Snapshot Should render correctly and match snapshot. 1`] = `
<div>
  <div
    class="css-cwg9oi-MuiStack-root"
  >
    <div
      class="css-le9p6c-MuiStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-body1 css-1ymyiii-MuiTypography-root"
      >
        Welcome to Minesweeper!
      </h1>
      <div
        class="css-j3eba0-MuiStack-root"
      >
        <div
          class="css-w004l2-MuiStack-root"
        >
          <legend
            class="MuiTypography-root MuiTypography-body1 css-8m5mf3-MuiTypography-root"
          >
            Difficulty
          </legend>
          <span
            class="MuiRating-root MuiRating-sizeLarge css-jue3ft-MuiRating-root"
          >
            <label
              class="css-dqr9h-MuiRating-label"
              for="mui-1"
            >
              <span
                class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
              >
                <svg
                  color="crimson"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: crimson;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M440.5 88.5l-52 52L415 167c9.4 9.4 9.4 24.6 0 33.9l-17.4 17.4c11.8 26.1 18.4 55.1 18.4 85.6 0 114.9-93.1 208-208 208S0 418.9 0 304 93.1 96 208 96c30.5 0 59.5 6.6 85.6 18.4L311 97c9.4-9.4 24.6-9.4 33.9 0l26.5 26.5 52-52 17.1 17zM500 60h-24c-6.6 0-12 5.4-12 12s5.4 12 12 12h24c6.6 0 12-5.4 12-12s-5.4-12-12-12zM440 0c-6.6 0-12 5.4-12 12v24c0 6.6 5.4 12 12 12s12-5.4 12-12V12c0-6.6-5.4-12-12-12zm33.9 55l17-17c4.7-4.7 4.7-12.3 0-17-4.7-4.7-12.3-4.7-17 0l-17 17c-4.7 4.7-4.7 12.3 0 17 4.8 4.7 12.4 4.7 17 0zm-67.8 0c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17zm67.8 34c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17zM112 272c0-35.3 28.7-64 64-64 8.8 0 16-7.2 16-16s-7.2-16-16-16c-52.9 0-96 43.1-96 96 0 8.8 7.2 16 16 16s16-7.2 16-16z"
                  />
                </svg>
              </span>
              <span
                class="MuiRating-visuallyHidden"
              >
                1 Star
              </span>
            </label>
            <input
              checked=""
              class="MuiRating-visuallyHidden"
              id="mui-1"
              name="difficulty"
              type="radio"
              value="1"
            />
            <label
              class="css-dqr9h-MuiRating-label"
              for="mui-2"
            >
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  color="white"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: white;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M440.5 88.5l-52 52L415 167c9.4 9.4 9.4 24.6 0 33.9l-17.4 17.4c11.8 26.1 18.4 55.1 18.4 85.6 0 114.9-93.1 208-208 208S0 418.9 0 304 93.1 96 208 96c30.5 0 59.5 6.6 85.6 18.4L311 97c9.4-9.4 24.6-9.4 33.9 0l26.5 26.5 52-52 17.1 17zM500 60h-24c-6.6 0-12 5.4-12 12s5.4 12 12 12h24c6.6 0 12-5.4 12-12s-5.4-12-12-12zM440 0c-6.6 0-12 5.4-12 12v24c0 6.6 5.4 12 12 12s12-5.4 12-12V12c0-6.6-5.4-12-12-12zm33.9 55l17-17c4.7-4.7 4.7-12.3 0-17-4.7-4.7-12.3-4.7-17 0l-17 17c-4.7 4.7-4.7 12.3 0 17 4.8 4.7 12.4 4.7 17 0zm-67.8 0c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17zm67.8 34c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17zM112 272c0-35.3 28.7-64 64-64 8.8 0 16-7.2 16-16s-7.2-16-16-16c-52.9 0-96 43.1-96 96 0 8.8 7.2 16 16 16s16-7.2 16-16z"
                  />
                </svg>
              </span>
              <span
                class="MuiRating-visuallyHidden"
              >
                2 Stars
              </span>
            </label>
            <input
              class="MuiRating-visuallyHidden"
              id="mui-2"
              name="difficulty"
              type="radio"
              value="2"
            />
            <label
              class="css-dqr9h-MuiRating-label"
              for="mui-3"
            >
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  color="white"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: white;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M440.5 88.5l-52 52L415 167c9.4 9.4 9.4 24.6 0 33.9l-17.4 17.4c11.8 26.1 18.4 55.1 18.4 85.6 0 114.9-93.1 208-208 208S0 418.9 0 304 93.1 96 208 96c30.5 0 59.5 6.6 85.6 18.4L311 97c9.4-9.4 24.6-9.4 33.9 0l26.5 26.5 52-52 17.1 17zM500 60h-24c-6.6 0-12 5.4-12 12s5.4 12 12 12h24c6.6 0 12-5.4 12-12s-5.4-12-12-12zM440 0c-6.6 0-12 5.4-12 12v24c0 6.6 5.4 12 12 12s12-5.4 12-12V12c0-6.6-5.4-12-12-12zm33.9 55l17-17c4.7-4.7 4.7-12.3 0-17-4.7-4.7-12.3-4.7-17 0l-17 17c-4.7 4.7-4.7 12.3 0 17 4.8 4.7 12.4 4.7 17 0zm-67.8 0c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17zm67.8 34c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17zM112 272c0-35.3 28.7-64 64-64 8.8 0 16-7.2 16-16s-7.2-16-16-16c-52.9 0-96 43.1-96 96 0 8.8 7.2 16 16 16s16-7.2 16-16z"
                  />
                </svg>
              </span>
              <span
                class="MuiRating-visuallyHidden"
              >
                3 Stars
              </span>
            </label>
            <input
              class="MuiRating-visuallyHidden"
              id="mui-3"
              name="difficulty"
              type="radio"
              value="3"
            />
            <label
              class="css-dqr9h-MuiRating-label"
              for="mui-4"
            >
              <span
                class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
              >
                <svg
                  color="white"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: white;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M440.5 88.5l-52 52L415 167c9.4 9.4 9.4 24.6 0 33.9l-17.4 17.4c11.8 26.1 18.4 55.1 18.4 85.6 0 114.9-93.1 208-208 208S0 418.9 0 304 93.1 96 208 96c30.5 0 59.5 6.6 85.6 18.4L311 97c9.4-9.4 24.6-9.4 33.9 0l26.5 26.5 52-52 17.1 17zM500 60h-24c-6.6 0-12 5.4-12 12s5.4 12 12 12h24c6.6 0 12-5.4 12-12s-5.4-12-12-12zM440 0c-6.6 0-12 5.4-12 12v24c0 6.6 5.4 12 12 12s12-5.4 12-12V12c0-6.6-5.4-12-12-12zm33.9 55l17-17c4.7-4.7 4.7-12.3 0-17-4.7-4.7-12.3-4.7-17 0l-17 17c-4.7 4.7-4.7 12.3 0 17 4.8 4.7 12.4 4.7 17 0zm-67.8 0c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17zm67.8 34c-4.7-4.7-12.3-4.7-17 0-4.7 4.7-4.7 12.3 0 17l17 17c4.7 4.7 12.3 4.7 17 0 4.7-4.7 4.7-12.3 0-17l-17-17zM112 272c0-35.3 28.7-64 64-64 8.8 0 16-7.2 16-16s-7.2-16-16-16c-52.9 0-96 43.1-96 96 0 8.8 7.2 16 16 16s16-7.2 16-16z"
                  />
                </svg>
              </span>
              <span
                class="MuiRating-visuallyHidden"
              >
                4 Stars
              </span>
            </label>
            <input
              class="MuiRating-visuallyHidden"
              id="mui-4"
              name="difficulty"
              type="radio"
              value="4"
            />
            <label
              class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
            >
              <input
                class="MuiRating-visuallyHidden"
                id="difficulty-empty"
                name="difficulty"
                type="radio"
                value=""
              />
              <span
                class="MuiRating-visuallyHidden"
              >
                Empty
              </span>
            </label>
          </span>
        </div>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButtonBase-root  css-1a9y98j-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Start
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="css-t9pp5n-MuiStack-root"
        data-testid="Board"
      >
        <div
          style="position: relative; height: 240px; width: 240px; overflow: auto; will-change: transform; direction: ltr;"
        >
          <div
            style="height: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
